File: question/part6_dynamicProgramming/src/edit_distance.cpp
#include "../../utils/common.hpp"

void printVector2D(const vector<vector<int>>& vec) {
    for (auto it = vec.begin(); it != vec.end(); ++it) {
        // it 为一个指针， 指向二维向量vec的一级迭代器
        for (auto jt = it->begin(); jt != it->end(); ++jt) {
            cout << *jt << " ";
        }
        cout << endl;
    }

    // for (vector<vector<int>>::iterator it = dp.begin(); it != dp.end(); it++){
    //     for (vector<int>::iterator jt = it->begin(); jt != it->end(); jt++){
    //         cout << *jt << " ";
    //     }
    //     cout << endl;
    // }
}


int edit_distance(string s1, string s2){
    if (s1.size() == 0){
        return s2.size();
    }
    if (s2.size() == 0){
        return s1.size();
    }

    vector<vector<int>>dp(s1.size()+1, vector<int>(s2.size()+1, 0));
    for (int i=0; i<s1.size()+1; i++){
        dp[i][0] = i;
    }
    for (int j=0; j<s2.size()+1; j++){
        dp[0][j] = j;
    }

    for (int i=1; i<s1.size()+1; i++){
        for (int j=1; j<s2.size()+1; j++){
            if (s1[i-1] == s2[j-1]){
                dp[i][j] = dp[i-1][j-1];
            }
            else{
                dp[i][j] = min(min(dp[i-1][j], dp[i][j-1]), dp[i-1][j-1]) + 1;
            }
        }
    }  
    return dp[s1.size()][s2.size()];
}

int main(){
    cout << edit_distance("bag", "pack") << endl; 
    return 0;
}

File: question/part6_dynamicProgramming/script/edit_distance.py
"""
输入两个字符串 $s$ 和 $t$ ，返回将 $s$ 转换为 $t$ 所需的最少编辑步数。
你可以在一个字符串中进行三种编辑操作：插入一个字符、删除一个字符、替换字符为任意一个字符
"""

def edit_distance(s1: str, s2: str) -> int:
    """
    动态规划实现编辑距离
    :param s1:
    :param s2:
    :return:
    """
    if len(s1) == 0:
        return len(s2)
    if len(s2) == 0:
        return len(s1)

    # 1. 定义 dp[i][j] 表示将s1的前i个字符更改为s2的前j个字符所需的最少编辑步数
    dp = [[0] * (len(s2) + 1) for _ in range(len(s1) + 1)]  
    # 2. 边界条件
    for j in range(len(s2)+1):
        dp[0][j] = j
    for i in range(len(s1)+1):
        dp[i][0] = i

    for i in range(1, len(s1)+1):
        for j in range(1, len(s2)+1):
            # 3. 转移方程
            if s1[i-1] == s2[j-1]:
                dp[i][j] = dp[i-1][j-1]
            else:
                dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + 1
    return dp[-1][-1]


if __name__ == '__main__':
    print(edit_distance("bag", "pack"))